### PR TITLE
#97 token refresh

### PR DESCRIPTION
**Pull request by ArdeaSt**
Implemented refresh token Authentication. Created Entity RefreshToken. Updated DB with new changes. Created TokenService and TokenController.  Made changes in current code for implementation refresh token model.

### CODE REVIEW

File: EventsExpress.Db/DbInitialize/DbInitializer.cs [hunk 1]
```diff
@@ -62,6 +62,10 @@ public static class DbInitializer
                 new Category{ Name="Team-Building"},
                 new Category{ Name="Swimming"},
                 new Category{ Name="Gaming"},
+                new Category {Name = "Fishing"},
```
Comment (sand0) [inline EventsExpress.Db/DbInitialize/DbInitializer.cs]: It would be better to use the same code stile

File: EventsExpress.Db/Entities/RefreshToken.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+
+namespace EventsExpress.Db.Entities
+{
+  public  class RefreshToken
+    {
+        public int Id { get; set; }
+        public string Token { get; set; }
+        public DateTime Expires { get; set; }
+        public bool IsExpired => DateTime.UtcNow >= Expires;
```
Comment (sand0) [inline EventsExpress.Db/Entities/RefreshToken.cs]: Do we need have any logic in EF entity model? On my mind this should be moved to Core layer DTO model

File: EventsExpress/ClientApp/src/services/EventsExpressService.js [hunk 3]
```diff
@@ -289,18 +290,54 @@ export default class EventsExpressService {
...
+
+            }
+            else {
+                return {
+                    error: {
+                        ErrorCode: res.status,
+                        massage: await res.statusText
+                    }
+                }
+            }
```
Comment (sand0) [inline EventsExpress/ClientApp/src/services/EventsExpressService.js]: use 
if () { 
 ... 
} else if() { 
 ... 
} else {
 ... 
}

File: EventsExpress/ClientApp/src/services/EventsExpressService.js [hunk 4]
```diff
@@ -371,26 +408,91 @@ export default class EventsExpressService {
...
+                    method: "POST"
+                });
+
+                if (!setResWithDataUpdateToken.ok) {
+                    return this.getResourceResponseHandler(setResWithDataUpdateToken);
+                }
+                let result = await this.getResourceResponseHandler(setResWithDataUpdateToken);
+                localStorage.setItem('token', result.jwtToken);
+
+                return this.setResourceWithData(url, data);
```
Comment (sand0) [inline EventsExpress/ClientApp/src/services/EventsExpressService.js]: DRY !!!

This should be moved into separate method.
You can use callback parameter for last return call

File: EventsExpress/EventsExpress.csproj [hunk 5]
```diff
@@ -11,7 +11,7 @@
   </PropertyGroup>
 
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
-    <DocumentationFile>EventsExpress.xml</DocumentationFile>
+    <DocumentationFile>C:\Users\Vlad Ymyr\Source\Repos\EventsExpress\EventsExpress\EventsExpress.xml</DocumentationFile>
```
Comment (sand0) [inline EventsExpress/EventsExpress.csproj]: Revert this change

File: EventsExpress.Db/EF/AppDbContextFactory.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+{
+    /// <summary>
+    /// ContextFactory needs for migration DB
+    /// </summary>
+    public class AppDbContextFactory : IDesignTimeDbContextFactory<AppDbContext>
+    {
+        public AppDbContext CreateDbContext(string[] args)
+        {
+            var optionsBuilder = new DbContextOptionsBuilder<AppDbContext>();
+            optionsBuilder.UseSqlServer("Server=(localdb)\\mssqllocaldb;Database=EventsExpress;Trusted_Connection=True;MultipleActiveResultSets=true");
```
Comment (sand0) [inline EventsExpress.Db/EF/AppDbContextFactory.cs]: Can we get this from appsettings.json? Hardcoded connection will cause error when migrations apply for test/release

File: EventsExpress.Core/IServices/IUsersService.cs [hunk 7]
```diff
@@ -32,6 +32,7 @@ public interface IUserService
         ProfileDTO GetProfileById(Guid id, Guid fromId);
 
         IEnumerable<UserDTO> Get(UsersFilterViewModel model, out int count, Guid id);
+        IEnumerable<UserDTO> GetAllUsers(string role1, string role2);
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/IServices/IUsersService.cs]: Please do not name parameters as `role1`, `role2`. What if we need more than two roles?
1. Consider using a parameter as a collection of roles.
2. Do you really need to get *all* users?

File: EventsExpress.Core/Services/AuthService.cs [hunk 8]
```diff
@@ -12,83 +12,98 @@
...
+            var jwtToken = _tokenService.GenerateAccessToken(user);
+            var refreshToken = _tokenService.GenerateRefreshToken();
 
-            var token = GenerateJwt(user);
+            // save refresh token
+            user.RefreshTokens = new List<RefreshToken> { refreshToken };
+            await _userService.Update(user);
 
-            return new OperationResult(true, token, "");
+            return (new OperationResult(true, jwtToken, ""), new AuthenticateResponseModel(jwtToken, refreshToken.Token));
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/AuthService.cs]: You don't need to set `jwtToken` into `Message` property anymore.

File: EventsExpress.Core/Services/AuthService.cs [hunk 9]
```diff
@@ -12,83 +12,98 @@
...
+            var jwtToken = _tokenService.GenerateAccessToken(userDto);
+            var refreshToken = _tokenService.GenerateRefreshToken();
 
-            return new OperationResult(true, token, "");
+            // save refresh token
+            userDto.RefreshTokens = new List<RefreshToken> { refreshToken };
+
+            await _userService.Update(userDto);
+
+            return (new OperationResult(true, jwtToken, ""), new AuthenticateResponseModel(jwtToken, refreshToken.Token));
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/AuthService.cs]: Same comment as above.

File: EventsExpress.Core/Services/TokenService.cs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                {
+                    Token = Convert.ToBase64String(randomNumber),
+                    Expires = DateTime.Now.AddDays(7),
+                    Created = DateTime.Now
+                };
+            }
+        }
+        public ClaimsPrincipal GetPrincipalFromExpiredToken(string token)
+        {
+            var signingKey = new SigningSymmetricKey(_configuration.GetValue<string>("JWTOptions:SecretKey"));
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/TokenService.cs]: Consider injecting `IOptions<T>` instead of getting configuration directly.

File: EventsExpress.Core/Services/TokenService.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            };
+            var tokenHandler = new JwtSecurityTokenHandler();
+            var principal = tokenHandler.ValidateToken(token, tokenValidationParameters, out SecurityToken securityToken);
+            if (!(securityToken is JwtSecurityToken jwtSecurityToken) || !jwtSecurityToken.Header.Alg.Equals(SecurityAlgorithms.HmacSha256, StringComparison.InvariantCultureIgnoreCase))
+                throw new SecurityTokenException("Invalid token");
+            return principal;
+        }
+        public async Task<AuthenticateResponseModel> RefreshToken(string token)
+        {
+           var user = _userService.GetAllUsers("User","Admin")
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/TokenService.cs]: You should add method to get user by refresh token into user service and not do in memory operations to improve performance. 

What if we add more roles in the future? This is not maintainable in such case.

File: EventsExpress.Core/Services/TokenService.cs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+            return new AuthenticateResponseModel(jwtToken, newRefreshToken.Token);
+        }
+        public void SetTokenCookie(string token)
+        {
+            var cookieOptions = new CookieOptions
+            {
+                HttpOnly = true,
+                Expires = DateTime.Now.AddDays(7)
+            };
+            MyHttpContext.Current.Response.Cookies.Delete("refreshToken");
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/TokenService.cs]: You should access `HttpContext` through `IHttpContextAccessor`, otherwise this code is not testable.

File: EventsExpress.Db/DbInitialize/DbInitializer.cs (hunk 1 above)
Comment (iuriikhrystiuk) [reply EventsExpress.Db/DbInitialize/DbInitializer.cs]: This task is being taken care of by Zhenya.

File: EventsExpress/ClientApp/src/services/EventsExpressService.js [hunk 13]
```diff
@@ -289,18 +290,54 @@ export default class EventsExpressService {
             method: "get",
             headers: new Headers({
                 'Content-Type': 'application/json',
-                'Authorization': `Bearer ${localStorage.getItem('token')}`
+                'Authorization': 'Bearer ' + localStorage.getItem('token')
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/EventsExpressService.js]: Consider better merging the code. String interpolation is superior in this case.

File: EventsExpress/ClientApp/src/services/EventsExpressService.js [hunk 14]
```diff
@@ -289,18 +290,54 @@ export default class EventsExpressService {
...
             }),
         });
 
-        return !res.ok
-            ? {
+        if (res.ok) {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/EventsExpressService.js]: Lines 297 through 326 are duplicated in three methods. Consider extracting those into separate method and reusing it.

File: EventsExpress/Controllers/AuthenticationController.cs [hunk 15]
```diff
@@ -41,27 +54,24 @@ IAuthService authSrv
...
+                {
+                return BadRequest(opResult.Message);
+                }
             var user = _userService.GetByEmail(authRequest.Email);
-
             var userInfo = _mapper.Map<UserInfo>(user);
-            userInfo.Token = result.Message;      
-
+            userInfo.Token = authResponseModel.JwtToken;
+            Response.Cookies.Delete("refreshToken", new CookieOptions { HttpOnly = true });
```
Comment (iuriikhrystiuk) [inline EventsExpress/Controllers/AuthenticationController.cs]: You are already deleting cookies here and in `SetTokenCookie` method. Pick one place.

File: EventsExpress/Controllers/AuthenticationController.cs [hunk 16]
```diff
@@ -135,11 +146,11 @@ public async Task<IActionResult> GoogleLogin([FromBody]UserView userView)
         public IActionResult Login()
         {
             var user = _authService.GetCurrentUser(HttpContext.User);
-            var userInfo = _mapper.Map<UserInfo>(user);
-
-            return Ok(userInfo);
+            return
+            user == null
+               ? (IActionResult)Unauthorized()
```
Comment (iuriikhrystiuk) [inline EventsExpress/Controllers/AuthenticationController.cs:189]: Is this cast really necessary? Please remove it if not.

File: EventsExpress/Controllers/AuthenticationController.cs [hunk 17]
```diff
@@ -181,23 +186,20 @@ public async Task<IActionResult> Register(LoginDto authRequest)
         public async Task<IActionResult> PasswordRecovery(string email)
         {
             if (string.IsNullOrEmpty(email))
-            {
+                {
```
Comment (iuriikhrystiuk) [inline EventsExpress/Controllers/AuthenticationController.cs]: Do not format this code in such way. Block statements are always on indentation of the conditional/loop statements. Statements inside blocks have indentation = block indentation + 1.

**Comment by iuriikhrystiuk** [comment]
I can see `Revoke` API method, but I cannot find when it is called. Is it supposed to be called during log out? Please either remove it, or implement client features for it.

**Review by iuriikhrystiuk — CHANGES_REQUESTED** [review]

File: EventsExpress.Core/IServices/IUsersService.cs (hunk 7 above)
Comment (ArdeaSt) [reply EventsExpress.Core/IServices/IUsersService.cs]: refactored

File: EventsExpress/EventsExpress.csproj (hunk 5 above)
Comment (ArdeaSt) [reply EventsExpress/EventsExpress.csproj]: did it

File: EventsExpress.Core/Services/AuthService.cs (hunk 9 above)
Comment (ArdeaSt) [reply EventsExpress.Core/Services/AuthService.cs]: deleted unuseful variables

File: EventsExpress.Core/Services/TokenService.cs (hunk 11 above)
Comment (ArdeaSt) [reply EventsExpress.Core/Services/TokenService.cs]: implemented GetUserByRefreshToken method on DB side for seaching user by refreshtoken

File: EventsExpress.Core/Services/TokenService.cs (hunk 12 above)
Comment (ArdeaSt) [reply EventsExpress.Core/Services/TokenService.cs]: changed code

File: EventsExpress/ClientApp/src/services/EventsExpressService.js (hunk 13 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/services/EventsExpressService.js]: changed

File: EventsExpress/Controllers/AuthenticationController.cs (hunk 15 above)
Comment (ArdeaSt) [reply EventsExpress/Controllers/AuthenticationController.cs]: Refacored, Deleting cookies in SetTokenCookie method.

File: EventsExpress/Controllers/AuthenticationController.cs (hunk 16 above)
Comment (ArdeaSt) [reply EventsExpress/Controllers/AuthenticationController.cs:189]: Yes its necessary , because there is no implicit conversion between '...BadRequestObjectResult' and '...OkResult'.

File: EventsExpress/Controllers/AuthenticationController.cs (hunk 17 above)
Comment (ArdeaSt) [reply EventsExpress/Controllers/AuthenticationController.cs]: refactored

File: EventsExpress.Db/Entities/RefreshToken.cs (hunk 2 above)
Comment (ArdeaSt) [reply EventsExpress.Db/Entities/RefreshToken.cs]: Created RefreshTokenDTO, and deleted method from RefreshTokenEntity

File: EventsExpress/ClientApp/src/services/EventsExpressService.js (hunk 14 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/services/EventsExpressService.js]: refactored

File: EventsExpress/ClientApp/src/services/EventsExpressService.js (hunk 4 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/services/EventsExpressService.js]: created functions RefreshHandler and ErrorHandler and used in this case

File: EventsExpress/ClientApp/src/services/EventsExpressService.js (hunk 3 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/services/EventsExpressService.js]: did it

File: EventsExpress.Db/Entities/RefreshToken.cs (hunk 2 above)
Comment (ArdeaSt) [reply EventsExpress.Db/Entities/RefreshToken.cs]: created RefreshTokenDTO and moved methods to this model

File: EventsExpress.Core/NotificationHandlers/RegisterVerificationHandler.cs [hunk 18]
```diff
@@ -58,11 +52,10 @@ public async Task Handle(RegisterVerificationMessage notification, CancellationT
 
                 var x = _cacheHepler.GetValue(notification.User.Id);
 
-                Debug.WriteLine(x.Token);
             }
-            catch(Exception ex)
+            catch
             {
-                Debug.WriteLine(ex.Message);
+
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/NotificationHandlers/RegisterVerificationHandler.cs]: Empty catch statements are very bad. You should always do something, otherwise why catching exceptions? Please consider injecting `ILogger<T>` interface and logging the exception explicitly, so we could debug and fix it if this happens in production.

File: EventsExpress.Core/Services/TokenService.cs (hunk 10 above)
Comment (iuriikhrystiuk) [reply EventsExpress.Core/Services/TokenService.cs]: Again, please use `IOptions<T>` - this has not been addressed.

File: EventsExpress.Core/IServices/ITokenService.cs [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public interface ITokenService
+    {
+        string GenerateAccessToken(UserDTO user);
+        RefreshToken GenerateRefreshToken(string ipAddress);
+        ClaimsPrincipal GetPrincipalFromExpiredToken(string token);
+        Task<AuthenticateResponseModel> RefreshToken(string token);
+        Task<bool> RevokeToken(string token);
+        void SetTokenCookie(string token);
+        string IpAddress();
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/IServices/ITokenService.cs]: Consider making this method `private` if it is not being used outside of the interface.

File: EventsExpress.Core/Services/UsersService.cs [hunk 20]
```diff
@@ -249,10 +251,20 @@ public IEnumerable<UserDTO> GetUsersByCategories(IEnumerable<CategoryDTO> catego
                     .Any(category => categoryIds.Contains(category.Category.Id)))
                 .Distinct()
                 .AsEnumerable();
-
+
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/UsersService.cs]: Please remove these redundant spaces. Consider fixing any warnings that you have in this PR.

File: EventsExpress/ClientApp/src/actions/logout.js [hunk 21]
```diff
@@ -4,17 +4,25 @@ import { reset_hub } from './chat';
...
 
   function setLogout() {
     return {
       type: SET_LOGOUT
     };
-  }
\ No newline at end of file
+}
+
+async function RevokeToken() {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/actions/logout.js]: Please follow naming convention on the UI and use camel-casing.

File: EventsExpress/ClientApp/src/actions/logout.js [hunk 22]
```diff
@@ -4,17 +4,25 @@ import { reset_hub } from './chat';
 import { resetNotification } from './chats';
 export const SET_LOGOUT = "SET_LOGOUT";
 
-  export default function logout(){
-    localStorage.clear();
-    return dispatch => { 
+export default  function logout() {
+       RevokeToken();
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/actions/logout.js]: The method is not awaited. Consider awaiting async methods.

File: EventsExpress/ClientApp/src/services/EventsExpressService.js [hunk 23]
```diff
@@ -293,14 +294,40 @@ export default class EventsExpressService {
...
+        }
+        else if (res.status === 401) {
+            return this.RefreshHandler(this.getResource, res, url);
+        }
+        else {
+            return this.errorHandler(res);
+        }
+    }
+
+    RefreshHandler = async (callback, res, url, data) => {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/EventsExpressService.js]: Same here - consider following standard naming convention for UI project.

File: EventsExpress/ClientApp/src/services/EventsExpressService.js [hunk 24]
```diff
@@ -293,14 +294,40 @@ export default class EventsExpressService {
...
-            ? {
-                error: {
-                    ErrorCode: res.status,
-                    massage: await res.statusText
-                }
+        if (res.ok) {
+            return await res.json();
+        }
+        else if (res.status === 401) {
+            return this.RefreshHandler(this.getResource, res, url);
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/EventsExpressService.js]: Is there a reason to use recursive calls? what would be the anchor of recursion in this case?

File: EventsExpress/EventsExpress.csproj [hunk 25]
```diff
@@ -1,92 +0,0 @@
-<Project Sdk="Microsoft.NET.Sdk.Web">
```
Comment (iuriikhrystiuk) [inline EventsExpress/EventsExpress.csproj]: Why was this proj file removed?

File: EventsExpress/Mapping/AutoMapperProfile.cs [hunk 26]
```diff
@@ -205,8 +213,15 @@ public AutoMapperProfile()
                 })));
 
             CreateMap<Message, MessageDto>().ReverseMap();
-                
+
+            #endregion
+
+            #region
```
Comment (iuriikhrystiuk) [inline EventsExpress/Mapping/AutoMapperProfile.cs]: Consider naming the region if you are using one.

**Review by iuriikhrystiuk — CHANGES_REQUESTED** [review]

File: EventsExpress/Mapping/AutoMapperProfile.cs (hunk 26 above)
Comment (ArdeaSt) [reply EventsExpress/Mapping/AutoMapperProfile.cs]: added name of region

File: EventsExpress/EventsExpress.csproj (hunk 25 above)
Comment (ArdeaSt) [reply EventsExpress/EventsExpress.csproj]: added this file and changed path for logging by default

File: EventsExpress/ClientApp/src/actions/logout.js (hunk 21 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/actions/logout.js]: renamed using camel-casing

File: EventsExpress.Core/Services/UsersService.cs (hunk 20 above)
Comment (ArdeaSt) [reply EventsExpress.Core/Services/UsersService.cs]: removed

File: EventsExpress.Core/IServices/ITokenService.cs (hunk 19 above)
Comment (ArdeaSt) [reply EventsExpress.Core/IServices/ITokenService.cs]: its method using in another authservice.

File: EventsExpress/Controllers/AuthenticationController.cs [hunk 27]
```diff
@@ -1,5 +1,6 @@
 ﻿using AutoMapper;
 using EventsExpress.Core.DTOs;
+using EventsExpress.Controllers;
```
Comment (sand0) [inline EventsExpress/Controllers/AuthenticationController.cs]: Why do we need this one?

File: EventsExpress/Controllers/AuthenticationController.cs [hunk 28]
```diff
@@ -41,27 +54,23 @@ IAuthService authSrv
...
+        public async Task<IActionResult> Login(LoginDto authRequest)
         {
             if (!ModelState.IsValid)
             {
                 return BadRequest(ModelState);
             }
-
-            var result = _authService.Authenticate(authRequest.Email, authRequest.Password);
-            if (!result.Successed)
+            var (opResult, authResponseModel) = await _authService.Authenticate(authRequest.Email, authRequest.Password);
```
Comment (sand0) [inline EventsExpress/Controllers/AuthenticationController.cs:67]: Just FYI - we can avoid tuples if AuthResponce will inherit OperationDetails

File: EventsExpress.Core/Services/TokenService.cs [hunk 29]
```diff
@@ -0,0 +1,153 @@
...
+            var jwtToken = new JwtSecurityToken(
+                claims: claims,
+                expires: DateTime.Now.AddSeconds(lifeTime),
+                signingCredentials: new SigningCredentials(
+                        _signingEncodingKey.GetKey(),
+                        _signingEncodingKey.SigningAlgorithm)
+            );
+            return new JwtSecurityTokenHandler().WriteToken(jwtToken);
+        }
+        public RefreshToken GenerateRefreshToken(string ipAddress)
```
Comment (sand0) [inline EventsExpress.Core/Services/TokenService.cs]: 1. You do not need to pass IP here - you can retrieve it by call this.IPAddress
2. It would be better to have smth like "public RefreshToken GenerateRefreshToken(UserDTO user)" as for JWT, so you do not need remember about save token logic outside TokenService

File: EventsExpress.Core/Services/TokenService.cs [hunk 30]
```diff
@@ -0,0 +1,153 @@
...
+                return new RefreshToken
+                {
+                    Token = Convert.ToBase64String(randomNumber),
+                    Expires = DateTime.Now.AddDays(7),
+                    Created = DateTime.Now,
+                    CreatedByIp = ipAddress
+                };
+            }
+        }
+        public ClaimsPrincipal GetPrincipalFromExpiredToken(string token)
```
Comment (sand0) [inline EventsExpress.Core/Services/TokenService.cs]: Shouldn't this be "GetPrincipalFromJWT" ?

File: EventsExpress.Core/Services/TokenService.cs [hunk 31]
```diff
@@ -0,0 +1,153 @@
...
+            var tokenHandler = new JwtSecurityTokenHandler();
+            var principal = tokenHandler.ValidateToken(token, tokenValidationParameters, out SecurityToken securityToken);
+            if (!(securityToken is JwtSecurityToken jwtSecurityToken) || !jwtSecurityToken.Header.Alg.Equals(SecurityAlgorithms.HmacSha256, StringComparison.InvariantCultureIgnoreCase))
+                throw new SecurityTokenException("Invalid token");
+            return principal;
+        }
+        public async Task<AuthenticateResponseModel> RefreshToken(string token)
+        {
+            var user = _userService.GetUserByRefreshToken(token);
+            // return null if no user found with token
```
Comment (sand0) [inline EventsExpress.Core/Services/TokenService.cs:119]: I think it isn't good to return NULL here. We must be able to say more apparently : "Failed".
Consider possibility to inherit OperationdDetails

File: EventsExpress.Core/Services/UsersService.cs [hunk 32]
```diff
@@ -178,7 +178,7 @@ public UserDTO GetByEmail(string email)
 
             if (user != null)
             {
-                user.CanChangePassword = string.IsNullOrEmpty(user.PasswordHash) ? false : true;
+                user.CanChangePassword = !string.IsNullOrEmpty(user.PasswordHash);
```
Comment (sand0) [inline EventsExpress.Core/Services/UsersService.cs:181]: You can move this into mapper config

File: EventsExpress.Core/Services/UsersService.cs [hunk 33]
```diff
@@ -234,8 +234,10 @@ public IEnumerable<UserDTO> Get(UsersFilterViewModel model, out int count, Guid
         public IEnumerable<UserDTO> GetUsersByRole(string role)
         {
             var users = Db.UserRepository.Get("Role")
-                .Where(user => user.Role.Name == role)
-                .AsEnumerable();
+               .Where(user => user.Role.Name == role)
+               .Include(user => user.RefreshTokens)
```
Comment (sand0) [inline EventsExpress.Core/Services/UsersService.cs:239]: use Db.UserRepository.Get(includeProperties: "Role,RefreshTokens") instead  .Include(user => user.RefreshTokens)

File: EventsExpress.Db/Entities/RefreshToken.cs [hunk 34]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+
+namespace EventsExpress.Db.Entities
+{
+  public  class RefreshToken
+    {
+        public int Id { get; set; }
```
Comment (sand0) [inline EventsExpress.Db/Entities/RefreshToken.cs]: inherit BaseEntity to have Id automatically

File: EventsExpress/DTO/UserInfo.cs [hunk 35]
```diff
@@ -19,6 +19,7 @@ public class UserInfo
         public string Token { get; set; }
         public bool AfterEmailConfirmation { get; set; } = false;
         public IEnumerable<CategoryDto> Categories { get; set; }
+        public IEnumerable<RefreshToken> RefreshTokens { get; set; }
```
Comment (sand0) [inline EventsExpress/DTO/UserInfo.cs]: Do we need this on UI app? We have already set Refresh token into cookie

File: EventsExpress/Mapping/AutoMapperProfile.cs [hunk 36]
```diff
@@ -19,8 +19,12 @@ public AutoMapperProfile()
 
             #region USER MAPPING
             CreateMap<User, UserDTO>()
+                .ForMember(dest => dest.Id, opts => opts.MapFrom(src => src.Id))
+                .ForMember(dest => dest.Email, opts => opts.MapFrom(src => src.Email))
+                .ForMember(dest => dest.RoleId, opts => opts.MapFrom(src => src.RoleId))
```
Comment (sand0) [inline EventsExpress/Mapping/AutoMapperProfile.cs]: We don't need this. Properties with same names will be mapped by default

File: EventsExpress/Mapping/AutoMapperProfile.cs [hunk 37]
```diff
@@ -19,8 +19,12 @@ public AutoMapperProfile()
...
                 .ForMember(dest => dest.Categories, opts => opts.MapFrom(src => src.Categories))
-                .ForMember(dest => dest.Events, opts => opts.Ignore());
+                .ForMember(dest => dest.Events, opts => opts.Ignore())
+                .ForMember(dest => dest.RefreshTokens, opts => opts.MapFrom(src => src.RefreshTokens));
```
Comment (sand0) [inline EventsExpress/Mapping/AutoMapperProfile.cs:24]: The same as above

File: EventsExpress.Core/NotificationHandlers/RegisterVerificationHandler.cs (hunk 18 above)
Comment (sand0) [reply EventsExpress.Core/NotificationHandlers/RegisterVerificationHandler.cs]: We need to discuss it and create story for logging integration. I guess nothing related work done before

**Review by sand0 — CHANGES_REQUESTED** [review]

File: EventsExpress.Core/NotificationHandlers/RegisterVerificationHandler.cs (hunk 18 above)
Comment (iuriikhrystiuk) [reply EventsExpress.Core/NotificationHandlers/RegisterVerificationHandler.cs]: Well, the thing is, that `ILogger<T>` is a default Microsoft interface that is implemented by most .Net core logging frameworks. By default, without any loggers, it outputs into std out. This is enough for this usecase. We can create the story further on.

File: EventsExpress.Core/Services/TokenService.cs (hunk 10 above)
Comment (ArdeaSt) [reply EventsExpress.Core/Services/TokenService.cs]: created JwtOptionsModel and used IOption<JwtOptionsModel >

File: EventsExpress.Core/IServices/ITokenService.cs (hunk 19 above)
Comment (ArdeaSt) [reply EventsExpress.Core/IServices/ITokenService.cs]: refactored. Created private prop IPAddress

File: EventsExpress/ClientApp/src/actions/logout.js (hunk 22 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/actions/logout.js]: deleted async/await

File: EventsExpress/ClientApp/src/services/EventsExpressService.js (hunk 23 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/services/EventsExpressService.js]: refacored code. Create local func

File: EventsExpress/ClientApp/src/services/EventsExpressService.js (hunk 24 above)
Comment (ArdeaSt) [reply EventsExpress/ClientApp/src/services/EventsExpressService.js]: refactored

File: EventsExpress/Controllers/AuthenticationController.cs (hunk 27 above)
Comment (ArdeaSt) [reply EventsExpress/Controllers/AuthenticationController.cs]: deleted unuseful using

File: EventsExpress.Core/Services/TokenService.cs (hunk 30 above)
Comment (ArdeaSt) [reply EventsExpress.Core/Services/TokenService.cs]: renamed

File: EventsExpress/Mapping/AutoMapperProfile.cs (hunk 36 above)
Comment (ArdeaSt) [reply EventsExpress/Mapping/AutoMapperProfile.cs]: deleted

File: EventsExpress/DTO/UserInfo.cs (hunk 35 above)
Comment (ArdeaSt) [reply EventsExpress/DTO/UserInfo.cs]: deleted

File: EventsExpress.Db/Entities/RefreshToken.cs (hunk 34 above)
Comment (ArdeaSt) [reply EventsExpress.Db/Entities/RefreshToken.cs]: yes, but I think is changing nothing if I write this manually

File: EventsExpress.Db/EF/AppDbContextFactory.cs (hunk 6 above)
Comment (ArdeaSt) [reply EventsExpress.Db/EF/AppDbContextFactory.cs]: I i can't get  connStr from appsetting, because project EventExpressDb haven't reference  on EventExpress project

File: EventsExpress/Controllers/AuthenticationController.cs (hunk 28 above)
Comment (ArdeaSt) [reply EventsExpress/Controllers/AuthenticationController.cs:67]: ok, this time I leave here tuple.

File: EventsExpress.Core/Services/TokenService.cs (hunk 29 above)
Comment (ArdeaSt) [reply EventsExpress.Core/Services/TokenService.cs]: 1. did it

File: EventsExpress/Mapping/AutoMapperProfile.cs (hunk 37 above)
Comment (ArdeaSt) [reply EventsExpress/Mapping/AutoMapperProfile.cs:24]: deleted

File: EventsExpress.Core/IServices/ITokenService.cs [hunk 38]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public interface ITokenService
+    {
+        string GenerateAccessToken(UserDTO user);
+        RefreshToken GenerateRefreshToken();
+        ClaimsPrincipal GetPrincipalFromJwt(string token);
+        Task<AuthenticateResponseModel> RefreshToken(string token);
+        Task<bool> RevokeToken(string token);
+        void SetTokenCookie(string token);
+        //string IpAddress();
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/IServices/ITokenService.cs]: Remove this commented code.

File: EventsExpress.Core/Infrastructure/ConfigDbOptionsModel.cs [hunk 39]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Infrastructure/ConfigDbOptionsModel.cs]: Delete this file.

File: EventsExpress.Db/EF/AppDbContextFactory.cs (hunk 6 above)
Comment (iuriikhrystiuk) [reply EventsExpress.Db/EF/AppDbContextFactory.cs]: Just remove the file and use command parameters.

File: EventsExpress.Core/Services/TokenService.cs [hunk 40]
```diff
@@ -0,0 +1,158 @@
...
+
+        public async Task<bool> RevokeToken(string token)
+        {
+            var user = _userService.GetUserByRefreshToken(token);
+            if (user == null) return false;
+
+            var refreshToken = user.RefreshTokens.SingleOrDefault(x => x.Token == token);
+
+            // return false if token is not active
+            if (!_mapper.Map<RefreshTokenDTO>(refreshToken).IsActive || refreshToken == null) return false; ;
```
Comment (sand0) [inline EventsExpress.Core/Services/TokenService.cs]: double semicolons...

File: EventsExpress.Core/Services/UsersService.cs [hunk 41]
```diff
@@ -246,13 +248,23 @@ public IEnumerable<UserDTO> GetUsersByCategories(IEnumerable<CategoryDTO> catego
 
             var users = Db.UserRepository.Get("Photo,Role,Categories.Category")
                 .Where(user => user.Categories
-                    .Any(category => categoryIds.Contains(category.Category.Id)))
+                .Any(category => categoryIds.Contains(category.Category.Id)))
```
Comment (sand0) [inline EventsExpress.Core/Services/UsersService.cs:252]: This tabulation was correct - `Any()` related to `Categories`, not to `UserRepository.Get(...)`

**Review by iuriikhrystiuk — APPROVED** [review]

**Review by sand0 — APPROVED** [review]
